python/ml4ir/base/model/relevance_model.py

```diff
@@ -350,6 +361,27 @@ def from_univariate_interaction_model(
             logger=logger,
         )
 
+    def add_scheduler_as_callback(self):
+        """Adding reduce lr on plateau as a callback if specified"""
+        if self.model_config and 'lr_schedule' in self.model_config:
+            lr_schedule = self.model_config['lr_schedule']
+            lr_schedule_key = lr_schedule['key']
+            if lr_schedule_key == LearningRateScheduleKey.REDUCE_LR_ON_PLATEAU:
+                monitor_metric = self.monitor_metric
+                if monitor_metric is None:
+                    reduce_lr = tf.keras.callbacks.ReduceLROnPlateau(factor=lr_schedule.get('factor', 0.5),
+                                                                     patience=lr_schedule.get('patience', 1),
+                                                                     min_lr=lr_schedule.get('min_lr', 0.0001),
+                                                                     verbose=1)
+                elif not monitor_metric.startswith("val_"):
+                    monitor_metric = "val_{}".format(monitor_metric)
+                    reduce_lr = tf.keras.callbacks.ReduceLROnPlateau(monitor=monitor_metric,
+                                                                     factor=lr_schedule.get('factor', 0.5),
+                                                                     patience=lr_schedule.get('patience', 1),
+                                                                     min_lr=lr_schedule.get('min_lr', 0.0001),
+                                                                     verbose=1)
+                self.callbacks_list.append(reduce_lr)
```

Comment:
@mohazahran Move this code out of this module. You can have the callback addition here (down). The backend logic should be in a new folder, perhaps `callbacks/`